main/src/main/scala/sbt/ProjectMatrix.scala

```diff
@@ -0,0 +1,818 @@
+    def %(conf: String): ProjectMatrix.MatrixClasspathDependency =
+      ProjectMatrix.MatrixClasspathDependency(m, Some(conf))
+
+    def %(conf: Configuration): ProjectMatrix.MatrixClasspathDependency =
+      ProjectMatrix.MatrixClasspathDependency(m, Some(conf.name))
+  }
+
+  final case class MatrixClasspathDependency(
+      matrix: ProjectMatrixReference,
+      configuration: Option[String]
+  ) extends MatrixClasspathDep[ProjectMatrixReference]
+
+  private final class ProjectMatrixDef(
+      val id: String,
+      val base: File,
+      val scalaVersions: Seq[String],
+      val rows: Seq[ProjectRow],
+      val aggregate: Seq[ProjectMatrixReference],
+      val nonMatrixAggregate: Seq[ProjectReference],
+      val dependencies: Seq[MatrixClasspathDep[ProjectMatrixReference]],
+      val nonMatrixDependencies: Seq[ClasspathDep[ProjectReference]],
+      val settings: Seq[Def.Setting[_]],
+      val configurations: Seq[Configuration],
+      val plugins: Plugins,
+      val transforms: Seq[Project => Project],
+      val defAxes: Seq[VirtualAxis],
+  ) extends ProjectMatrix { self =>
+    lazy val resolvedMappings: ListMap[ProjectRow, Project] = resolveMappings
+    private def resolveProjectIds: Map[ProjectRow, String] = {
+      Map((for {
+        r <- rows
+      } yield {
+        val axes = r.axisValues
+          .sortBy(_.suffixOrder)
+          .filterNot(isSortOfDefaultAxis)
+        val idSuffix = axes.map(_.idSuffix).mkString("")
+        val childId = self.id + idSuffix
+        r -> childId
+      }): _*)
+    }
+
+    private def isSortOfDefaultAxis(a: VirtualAxis): Boolean =
+      defAxes exists { da => VirtualAxis.isPartialVersionEquals(da, a) }
+
+    private def resolveMappings: ListMap[ProjectRow, Project] = {
+      val projectIds = resolveProjectIds
+      def dirSuffix(axes: Seq[VirtualAxis]): String =
+        axes.map(_.directorySuffix).filter(_.nonEmpty).mkString("-")
+      val projects =
+        rows.map { r =>
+          import VirtualAxis.*
+          val axes = r.axisValues.sortBy(_.suffixOrder)
+          val scalaDirSuffix = dirSuffix(axes)
+          val nonScalaDirSuffix = dirSuffix(axes.filterNot(_.isInstanceOf[ScalaVersionAxis]))
+          val nonScalaNorPlatformDirSuffix =
+            dirSuffix(axes.filterNot(_.isInstanceOf[ScalaVersionAxis | PlatformAxis]))
+          val platform = axes
+            .collect { case pa: VirtualAxis.PlatformAxis =>
+              pa
+            }
+            .headOption
+            .getOrElse(sys.error(s"platform axis is missing in $axes"))
+          val childId = projectIds(r)
+          val deps = dependencies.map { resolveMatrixDependency(_, r) } ++ nonMatrixDependencies
+          val aggs = aggregate.map { case ref: LocalProjectMatrix =>
+            val other = lookupMatrix(ref)
+            resolveMatrixAggregate(other, r)
+          } ++ nonMatrixAggregate
+          val dotSbtMatrix = new java.io.File(".sbt") / "matrix"
```

Comment:
I wonder if we should change that path to `target/matrix`.

Reply:
Yea, something we can explore post merge I guess. Not sure how IDEs would behave if the base directories get wiped up.